src/connection/host_lookup.c

```diff
@@ -0,0 +1,541 @@
+      return gethostbyname_r (name, ret, buf, buflen, result, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+      return gethostbyname_r (name, ret, buf, buflen, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+      return gethostbyname_r (name, ret, &ht_data);
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif
+#endif
+    }
+  else
+    {
+      hp_buf = host_lookup_internal (name, NULL, HOSTNAME_TO_IPADDR);
+    }
+
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+  if (hp_buf == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return EFAULT;
+    }
+
+  if (((*result) = (struct hostent *) malloc (sizeof (struct hostent))) == NULL)
+    {
+      return NULL;
+    }
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+  memcpy ((void *) *result, (void *) hp_buf, sizeof (struct hostent));
+
+  return 0;
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+  if (hp_buf == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return NULL;
+    }
+
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+
+  return ret;
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+  if (hp_buf == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return -1;
+    }
+
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+
+  return 0;
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif /* HAVE_GETHOSTBYNAME_R */
+}
+
+int
+getnameinfo_uhost (struct sockaddr *addr, socklen_t addrlen, char *host, size_t hostlen, char *serv, size_t servlen,
+		   int flags)
+{
+  struct hostent *hp = NULL;
+
+  if (prm_get_bool_value (PRM_ID_USE_USER_HOSTS) == USE_GLIBC_HOSTS)
+    {
+      return getnameinfo (addr, addrlen, host, hostlen, NULL, 0, NI_NOFQDN);
+    }
+  else
+    {
+      if ((hp = host_lookup_internal (NULL, addr, IPADDR_TO_HOSTNAME)) != NULL)
+	{
+	  strncpy (host, hp->h_name, hostlen);
+	  host[CUB_MAXHOSTNAMELEN - 1] = '\0';
+
+	  return 0;
+	}
+
+      return EINVAL;
+    }
+
+}
+
+int
+getaddrinfo_uhost (char *node, char *service, struct addrinfo *hints, struct addrinfo **res)
+{
+  struct hostent *hp = NULL;
+  struct addrinfo results_out;
+  struct sockaddr_in addr_convert;
+  struct in_addr *in_addr_buf = NULL;
+
+  if (prm_get_bool_value (PRM_ID_USE_USER_HOSTS) == USE_GLIBC_HOSTS)
+    {
+      return getaddrinfo (node, service, hints, res);
+    }
+  else
+    {
+      hp = host_lookup_internal (node, NULL, HOSTNAME_TO_IPADDR);
+    }
+
+  if (hp == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return EAI_NODATA;
+    }
+
+  if ((in_addr_buf = (struct in_addr *) malloc (sizeof (struct in_addr))) == NULL)
+    {
+      free (hp);
```

Comment:
A memory leak may occur. (line 500, 509)

Reply:
Right. I can guess all of hp elements de- allocation is needed